### PR TITLE
Kraken for U-root

### PR DESCRIPTION
**Pull request by synackd**
Adds functionality to be able to build a u-root-embeddable version of Kraken.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ munge.key
 ssh_host_*key*
 **/.DS_Store
 /kraken-build
+**/*.swp
```
Comment (jlowellwofford) [inline .gitignore]: While this isn't a bad idea, it should probably be its own PR.

File: config/uefi.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# kraken-build.go: describes a build for a Virtualbox testbed
```
Comment (jlowellwofford) [inline config/uefi.yaml]: Fix the comment

**Comment by synackd** [comment]
Since rebasing, there is no longer a `vendor` directory. Therefore, I'm working on getting rid of changing the `GOPATH` before building and relying on Go modules.

File: config/uefi.yaml (hunk 2 above)
Comment (synackd) [reply config/uefi.yaml]: Fixed.

File: .gitignore (hunk 1 above)
Comment (synackd) [reply .gitignore]: See #156

**Comment by synackd** [comment]
Found some Kraken commits that caused my version in KrakenBoot to be out-of-date, so I rebase on top of the latest commit.

File: go.mod [hunk 3]
```diff
@@ -7,8 +7,10 @@ require (
 	github.com/gorilla/mux v1.7.3
 	github.com/gorilla/websocket v1.4.1
 	github.com/mdlayher/raw v0.0.0-20191009151244-50f2db8cc065
+	github.com/otiai10/copy v1.0.1
 	github.com/pin/tftp v2.1.0+incompatible
 	github.com/satori/go.uuid v1.2.0
+	golang.org/x/crypto v0.0.0-20191122220453-ac88ee75c92c
```
Comment (jlowellwofford) [inline go.mod]: run `go mod tidy` to get rid of unused packages

File: lib/buildutil.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Simple search and replace
```
Comment (jlowellwofford) [inline lib/buildutil.go]: Fix `// SimpleSearchAndReplace ` commenting

File: kraken-build.go [hunk 5]
```diff
@@ -114,6 +117,124 @@ func compileTemplates(krakenDir, tmpDir string) (targets []string, e error) {
...
+
+				targets = append(targets, target)
+			}
+		}
+	}
+	return
+}
+
+// Generate kraken source tree for u-root command from krakenDir into outDir
+func uKraken(outDir, krakenDir string) (targets []string, e error) {
```
Comment (jlowellwofford) [inline kraken-build.go]: Can we split this unto sub-functions?

File: go.mod (hunk 3 above)
Comment (synackd) [reply go.mod]: Committed.

File: kraken-build.go (hunk 5 above)
Comment (synackd) [reply kraken-build.go]: The complexity did not seem to be with `uKraken()`, but with `main()` after I added the `if` block checking if `-uroot` is set. After moving this block to a separate function `checkUrootOpt()`, the complexity issue goes away.

I did, however, simplify the large `if`-`else` block in `uKraken()` since there was some unnecessary repeated code.

File: go.mod [hunk 6]
```diff
@@ -7,14 +7,15 @@ require (
 	github.com/gorilla/mux v1.7.3
 	github.com/gorilla/websocket v1.4.1
 	github.com/mdlayher/raw v0.0.0-20191009151244-50f2db8cc065
+	github.com/otiai10/copy v1.0.1
```
[the remark this answers is not in the document]
Comment (synackd) [reply go.mod:10]: This is addressed [here](https://github.com/synackd/kraken/pull/1/files).

File: go.mod [hunk 7]
```diff
@@ -7,14 +7,15 @@ require (
...
-	golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e // indirect
-	golang.org/x/text v0.3.2 // indirect
-	google.golang.org/genproto v0.0.0-20191115221424-83cc0476cb11 // indirect
 	google.golang.org/grpc v1.25.1
 	gopkg.in/yaml.v2 v2.2.7
 )
 
 go 1.13
+
+replace github.com/hpc/kraken => ./
```
Comment (synackd) [inline go.mod:21]: This needs reversion before merging.

File: kraken-build.go [hunk 8]
```diff
@@ -114,14 +114,288 @@ func compileTemplates(krakenDir, tmpDir string) (targets []string, e error) {
...
+				}
+				var target string
+				if target, e = compileTemplate(filepath.Join(krakenDir, "kraken", f.Name()), tmpDir); e != nil {
+					return
+				}
+
+				// Use import paths of generated files for u-root instead of the main
+				// kraken ones
+				var from, to, targetPath string
+				from = "hpc/kraken/"
```
[the remark this answers is not in the document]
Comment (synackd) [reply kraken-build.go:140]: Addressed by #184.

File: go.mod (hunk 7 above)
Comment (synackd) [reply go.mod]: Done: be9d8c

File: kraken-build.go [hunk 9]
```diff
@@ -114,14 +114,288 @@ func compileTemplates(krakenDir, tmpDir string) (targets []string, e error) {
 	return
 }
 
-func buildKraken(dir string, fromTemplates []string, t Target, verbose bool) (e error) {
+// compileMainTemplates is like CompileMainTemplates, but does so for sources that
+// will end up as a u-root command
+func compileUrootTemplates(krakenDir, tmpDir string) (targets []string, e error) {
```
[the remark this answers is not in the document]
Comment (synackd) [reply kraken-build.go]: Done: be9d8c0

This was actually pretty straight forward. Let's see if it passes code factor...

File: examples/vbox/kraken.yml [hunk 10]
```diff
@@ -1,8 +1,8 @@
 ---
 - hosts: kraken
   vars:
-    kr_repo: https://github.com/hpc/kraken.git
-    kr_repo_version: "master"
+    kr_repo: https://github.com/synackd/kraken.git
```
[the remark this answers is not in the document]
Comment (synackd) [reply examples/vbox/kraken.yml]: Done: 54d5149

File: config/uefi.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# kraken-build.go: describes a build for embedding in UEFI firmware
+targets:
+  'u-root':
```
[the remark this answers is not in the document]
Comment (synackd) [reply config/uefi.yaml]: Done: 4025d5f

File: kraken-build.go [hunk 12]
```diff
@@ -114,14 +114,288 @@ func compileTemplates(krakenDir, tmpDir string) (targets []string, e error) {
...
+	if e = lib.RegexReplace(dir, from, to); e != nil {
+		e = fmt.Errorf("unable to replace import paths for files in %s: %v", dir, e)
+		return
+	}
+
+	// ServiceInstance.go: BusyBox cannot handle argv[0] not being kraken.
+	// TODO: Use one replace for both lines.
+	siPath := path.Join(dir, "core/ServiceInstance.go")
+	e = lib.SearchAndReplace(siPath,
+		"si.cmd.Args = []string{\"[kraken:\" + si.ID() + \"]\"}",
```
[the remark this answers is not in the document]
Comment (synackd) [reply kraken-build.go:287]: Addressed by #185.

File: config/vbox.yaml [hunk 13]
```diff
@@ -1,5 +1,6 @@
 # kraken-build.go: describes a build for a Virtualbox testbed
 targets:
+  'u-root':
```
[the remark this answers is not in the document]
Comment (synackd) [reply config/vbox.yaml:3]: We discussed possibly putting an output path here, though that would require generation of a `go.mod` so that a path outside `GOPATH` can be specified. What do you think?

[the remark this answers is not in the document]
Comment (synackd) [reply config/vbox.yaml:3]: So instead of having `os` and `arch`, have, e.g., `destination`. The default value could be `""` which would generate the tree into `build/u-root/`.

**Comment by synackd** [comment]
Closing #100 since this was merged.
